copy/copy.go

```diff
@@ -862,6 +869,14 @@ func (ic *imageCopier) copyLayers(ctx context.Context) error {
 		}()
 
 		for i, srcLayer := range srcInfos {
+			if srcLayer.Annotations == nil {
+				srcLayer.Annotations = make(map[string]string)
+			}
+			// Add default labels containing information of the target image and layers.
+			srcLayer.Annotations[layerTargetDigest] = srcLayer.Digest.String()
+			if dref := ic.c.rawSource.Reference().DockerReference(); dref != nil {
+				srcLayer.Annotations[layerTargetReference] = dref.String()
```

Comment:
- This exposes private information; it’s not expected that the destination of a copy is told about the name used at the source.
- Why the full Docker reference, anyway? c/storage keys things off an image ID, and a Docker reference can refer to many different images over time.

Reply:
Registry-backed additional layer store filesystem (e.g. eStargz, zstd:chunked) needs to query layer data to the registry. The registry API [requires image reference](https://docs.docker.com/registry/spec/api/#pulling-a-layer) so this needs to be passed to the c/storage's `Store` and the underlying filesystem somehow.

Reply:
Having now read a bit more, https://github.com/containers/storage/pull/795#discussion_r551319228 : this is just data that should be passed as parameters, not by modifying the manifest, which affects every single `skopeo copy` including things like inter-registry copies exposing locations of private staging registries when making a product release.

(Even given that, it’s not at all clear to to me how this is going to work with authenticated registries; the credentials used by `copy.Image` are not made available to the on-demand additional store implementations by this. But that’s a secondary concern and probably solvable at least for the simple cases.)

Reply:
could we use a lookaside cache for storing such information (something implemented with: https://github.com/containers/storage/pull/787)?

Reply:
What if 3 different images that share a layer are pulled, one after another; 2 from a single registry, the third from a completely different registry; the three pulls are using different credentials, and later _some_ of the credentials are invalidated? The credentials are conceptually a property of (end user, repository), not a layer.  (And the concept of “end user” gets rather tricky once CRI-O is involved.)

I suppose just starting with unauthenticated repositories would still be useful. Getting the complex cases right might be more interesting.

Reply:
> this is just data that should be passed as parameters, not by modifying the manifest, which affects every single skopeo copy including things like inter-registry copies exposing locations of private staging registries when making a product release.

Reference is now pased as parameters to `c/storage.Store`, through a new API `*c/image.storageImageDestination.TryReusingBlobWithRef()`.

Reply:
> Even given that, it’s not at all clear to to me how this is going to work with authenticated registries

As initial implementation, the underlying filesystem's authentication logic (creds management) is separated from `copy.Image`.
We can come up with the way to sync creds between them but this can be following-up PRs.

Reply:
> We can come up with the way to sync creds between them but this can be following-up PRs.

I can’t much imagine how that would look like; so is this “will never happen”? The tradeoffs are relevant for choosing an an approach/implementation.